### PR TITLE
#1119 data table new component props

### LINKED ISSUE #1119: New Data Table components require new props
## Is your feature request related to a problem? Please describe.

The new components added for the new `DataTable` - `Row`, `Cell` etc. require new props for styling.

## Describe the solution you'd like

a `prop` added for each styling required i.e. `viewStyle` to style the `View` in a component, or `textStyle` to style the `Text` component.

As it is a very common use case that components in react-native use `flex` to adjust their size, and that this is normally a dynamic property, I think it is a good idea to have components have a `width` optional property, which when used injects the passed style (or default) with a `flex` property with a value of the passed `width prop`.

Another common use case is to use `borderRight` in a row based table for indicating the column seperators. I think it would be beneficial to have a `isLastCell` property, which when true, will inject the style with a `borderRightWidth: 0` property.

## Describe alternatives you've considered

Using `context` and `defaultProps` extensively. I think the above solution of using `props` is cleaner and simpler

Not having the additional `props` and having this logic within a `Page` component which uses the components. I think the use-cases are extremely common and if written in a way which can be ignored is a great feature

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1119 

## Change summary

- Added `isLastCell` and `width` `props` to data table components as described in the issue.
- Added `style` props to each component
- Added a `getAadjustedStyle` method in the `utilities` folder of `DataTable` to inject `flex` and  `borderRightWidth: 0` properties into a cells style

## Testing

**Internal code not really testable?**

### Related areas to think about

Default props?

### CODE REVIEW

**Comment by joshxg** [comment]
Did some hacking to see if there was gonna be a quick fix.

#### Navigation Stack re-rendering
- `ScreenProps` in `ReduxNavigator` is causing the large re-renders. Whenever `mSupplyMobileAppContainer` re-renders, all components in the navigation stack also re-render. Would need to be set as a constant somewhere (`componentDidMount` most likely) rather than a new object.

Current screenProps:
```
    this.screenProps = {
      database: this.database,
      settings: this.settings,
      logOut: this.logOut,
      currentUser: this.state.currentUser,
      database: this.database,
      runWithLoadingIndicator: this.runWithLoadingIndicator,
      isInAdminMode: this.state.isInAdminMode,
    };
```

#### Causes of `mSupplyMobileAppContainer` re-rendering:
- `syncState` also causes `mSupplyMobileAppContainer` to re-render. Propose that `SyncModal` and `NavigationBar` subscribe to the redux store to fetch the `syncState`, rather than being passed by `mSupplyMobileAppContainer`.
- `isLoading` causes `mSupplyMobileAppContainer` to re-render. It needs to in order to display the spinner, but fixing the above problem with `screenProps` fixes this.

#### Causes of `ReduxNavigation` re-render
- `ReduxNavigation` is being passed the current `navigationState` by `mSupplyMobileAppContainer`, which it does not use (but is always unequal) and also through redux in props also, causing every re-rendering to also re-render `ReduxNavigation`.
- Should be/can be `memoized`

Overall there doesn't seem to be a quick fix for the whole problem, but there are some which can be easily improved. I think that it is important to refactor this component. With a few hacks I improved the layout time of navigating to certain components by a lot - from ~900ms layout time to around ~150ms

File: src/widgets/DataTable/Cell.js [hunk 1]
```diff
@@ -1,36 +1,48 @@
...
+import { getAdjustedStyle } from './utilities';
 
 /**
  * Renders a cell that displays a string/number value within
  * a react-native `Text` component.
  *
- * @param {string|number} value The value to render in cell
+ * @param {string|number} value     The value to render in cell
+ * @param {Object}        viewStyle Style object for the containing View
+ * @param {Object}        textStyle Style object for the inner Text
```
Comment (Chris-Petty) [inline src/widgets/DataTable/Cell.js]: Add `isLastCell`

File: src/widgets/DataTable/Cell.js [hunk 2]
```diff
@@ -1,36 +1,48 @@
...
+  width: PropTypes.number,
+  isLastCell: PropTypes.bool,
 };
 
 Cell.defaultProps = {
   value: '',
+  textStyle: defaultStyles.textStyle,
+  viewStyle: defaultStyles.viewStyle,
+  width: 0,
+  isLastCell: 0,
```
Comment (Chris-Petty) [inline src/widgets/DataTable/Cell.js]:
```suggestion
  isLastCell: false,
```

File: src/widgets/DataTable/Cell.js [hunk 3]
```diff
@@ -1,36 +1,48 @@
...
+    <View style={internalViewStyle}>
+      <Text style={textStyle}>{value}</Text>
     </View>
   );
 });
 
+const defaultStyles = StyleSheet.create({
+  viewStyle: {},
+  textStyle: {},
+});
```
Comment (Chris-Petty) [inline src/widgets/DataTable/Cell.js]: Seems kind of redundant?

Comment (Chris-Petty) [reply src/widgets/DataTable/Cell.js]: Suppose it makes the old just handing an array of styles with overrides work the same?

File: src/widgets/DataTable/DataTable.js [hunk 4]
```diff
@@ -7,31 +8,41 @@ import PropTypes from 'prop-types';
 import React from 'react';
 import { StyleSheet, VirtualizedList, VirtualizedListPropTypes } from 'react-native';
 
-const DataTable = React.memo(({ renderRow, renderHeader, ...otherProps }) => (
+/**
+ * Base DataTable component. Thin wrapper around VirtualizedList, providing
+ * a header component. All VirtualizedList props can be passed through,
+ * however renderItem is renamed renderRow.
+ *
+ * @param {Func}   renderRow    renaming of virtualizedList renderItem prop.
```
Comment (Chris-Petty) [inline src/widgets/DataTable/DataTable.js]:
```suggestion
 * @param {Func}   renderRow    Renaming of VirtualizedList renderItem prop.
```

File: src/widgets/DataTable/HeaderCell.js [hunk 5]
```diff
@@ -14,6 +16,8 @@ import { TouchableOpacity, Text, StyleSheet, View, TouchableOpacityPropTypes } f
  * @prop {ReactElement} SortAscComponent      Icon component for ascending sorting
  * @prop {ReactElement} SortDescComponent     Icon component for descending sorting
  * @prop {ReactElement} SortNeutralComponent  Icon component for neutral state, no sort.
+ * @prop {Object}       containerStyle        Style object for the wrapping Touchable or View.
+ * @prop {Object}       textStyle             Style object for the inner text component.
```
Comment (Chris-Petty) [inline src/widgets/DataTable/HeaderCell.js]: Hope I'm not too annoying
```suggestion
 * @prop {Object}       textStyle             Style object for the inner Text component.
```

File: src/widgets/DataTable/HeaderCell.js [hunk 6]
```diff
@@ -66,6 +82,10 @@ HeaderCell.propTypes = {
   SortAscComponent: PropTypes.element,
   SortDescComponent: PropTypes.element,
   SortNeutralComponent: PropTypes.element,
+  containerStyle: PropTypes.object,
+  textStyle: PropTypes.object,
+  width: PropTypes.number,
+  sortable: PropTypes.bool,
```
Comment (Chris-Petty) [inline src/widgets/DataTable/HeaderCell.js]: `isLastCell` missing

File: src/widgets/DataTable/HeaderCell.js [hunk 7]
```diff
@@ -1,7 +1,9 @@
+/* eslint-disable react/forbid-prop-types */
```
Comment (Chris-Petty) [inline src/widgets/DataTable/HeaderCell.js]: I reckon we just turn this off globally

File: src/widgets/DataTable/TouchableCell.js [hunk 8]
```diff
@@ -52,18 +69,16 @@ TouchableCell.propTypes = {
   dispatch: PropTypes.func.isRequired,
   renderChildren: PropTypes.func.isRequired,
   TouchableComponent: PropTypes.func.isRequired,
+  containerStyle: PropTypes.object,
+  textStyle: PropTypes.object,
+  isLastCell: PropTypes.bool,
```
Comment (Chris-Petty) [inline src/widgets/DataTable/TouchableCell.js]: missing `width`

File: src/widgets/DataTable/utilities.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+/* eslint-disable import/prefer-default-export */
+export const getAdjustedStyle = (style, width, isLastCell) => {
```
Comment (Chris-Petty) [inline src/widgets/DataTable/utilities.js]: JS doc?

File: src/widgets/DataTable/Cell.js (hunk 1 above)
Comment (joshxg) [reply src/widgets/DataTable/Cell.js]: Oops missed a few added here f51041edef541cdc14ec5d1cfb8ac21dd374992f

File: src/widgets/DataTable/Cell.js (hunk 3 above)
Comment (joshxg) [reply src/widgets/DataTable/Cell.js]: I definitely don't like this, but I'm not sure what to put. I was thinking of adding either our styles, or some basic styles for it to still look like a table before porting them over to `react-native-data-table`?

File: src/widgets/DataTable/EditableCell.js [hunk 10]
```diff
@@ -27,6 +27,9 @@ import { getAdjustedStyle } from './utilities';
  * @param {Object} touchableStyle Style object for the wrapping Touchable component
  * @param {Object} viewStyle Style object for the wrapping View component
  * @param {Object} textStyle Style object for the inner Text component
+ * @param {Number}        width Optional flex property to inject into styles.
+ * @param {Bool}          isLastCell Indicator for if this cell is the last
```
Comment (Chris-Petty) [inline src/widgets/DataTable/EditableCell.js]: This why I don't bother aligning white space generally. I'll generally start the comment on tab after the param name and thats it.

File: src/widgets/DataTable/HeaderCell.js [hunk 11]
```diff
@@ -9,15 +9,18 @@ import { getAdjustedStyle } from './utilities';
...
+ * @param {String}       columnKey             The key for the column the cell heads.
+ * @param {Func}         onPressAction         Action for dispatching on press
+ * @param {Func}         dispatch              Dispatcher to backing reducer
+ * @param {ReactElement} SortAscComponent      Icon component for ascending sorting
+ * @param {ReactElement} SortDescComponent     Icon component for descending sorting
+ * @param {ReactElement} SortNeutralComponent  Icon component for neutral state, no sort.
+ * @param {Object}       containerStyle        Style object for the wrapping Touchable or View.
+ * @param {Object}       textStyle             Style object for the inner text component.
+ * @param {Number}        width               Optional flex property to inject into styles.
+ * @param {Bool}          isLastCell          Indicator for if this cell is the last
```
Comment (Chris-Petty) [inline src/widgets/DataTable/HeaderCell.js]: Once space out 😛

File: src/widgets/DataTable/HeaderCell.js (hunk 5 above)
Comment (joshxg) [reply src/widgets/DataTable/HeaderCell.js]: 587bfcd3594d21a1a3a206d05cdcc7c1d1499773

Note to self: commit all the suggestions first :(

File: src/widgets/DataTable/HeaderCell.js (hunk 6 above)
Comment (joshxg) [reply src/widgets/DataTable/HeaderCell.js]: dbf837f7e9dcd5894189d2b535575dd2daa9a3a5

File: src/widgets/DataTable/HeaderCell.js (hunk 7 above)
Comment (joshxg) [reply src/widgets/DataTable/HeaderCell.js]: I'll do a PR into develop, I reckon?

File: src/widgets/DataTable/TouchableCell.js (hunk 8 above)
Comment (joshxg) [reply src/widgets/DataTable/TouchableCell.js]: 57ef234e0b7c00a885b67abf201ed88a8aedf4e9

File: src/widgets/DataTable/utilities.js (hunk 9 above)
Comment (joshxg) [reply src/widgets/DataTable/utilities.js]: 5030ce17dc5cc74477a90aa295569ee00ea0f881

File: src/widgets/DataTable/HeaderCell.js (hunk 11 above)
Comment (joshxg) [reply src/widgets/DataTable/HeaderCell.js]: I do not have the required skills to fix this :(

**Comment by joshxg** [comment]
Have removed default styles from components

**Pull request merged**